### PR TITLE
HDDS-6131: EC: Replication config from bucket should be refreshed in o3fs

### PR DESCRIPTION
**Pull request by umamaheswararao**
## What changes were proposed in this pull request?

Client side we are giving bucket defaults preference in case if bucket has non ec configs. 
In other cases, we will determine the client side configs either from configured value or API passed value.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6131

## How was this patch tested?

Running existing test and will add tests

### CODE REVIEW

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneClientAdapterImpl.java [hunk 1]
```diff
@@ -211,29 +239,20 @@ public OzoneFSOutputStream createFile(String key, short replication,
...
+      OzoneOutputStream ozoneOutputStream = bucket.createFile(key, 0,
+          OzoneClientUtils.resolveClientSideReplicationConfig(replication,
+              clientConfiguredReplConfig, bucketReplConfig, config), overWrite,
+          recursive);
+
       return new OzoneFSOutputStream(ozoneOutputStream.getOutputStream());
     } catch (OMException ex) {
-      if (ex.getResult() == OMException.ResultCodes.FILE_ALREADY_EXISTS
-          || ex.getResult() == OMException.ResultCodes.NOT_A_FILE) {
+      if (ex.getResult() == OMException.ResultCodes.FILE_ALREADY_EXISTS || ex
```
Comment (fapifta) [inline hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneClientAdapterImpl.java:254]: This seems to be an accidental reformat, can we revert this please?

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneClientAdapterImpl.java [hunk 2]
```diff
@@ -407,6 +437,11 @@ public String getCanonicalServiceName() {
     return objectStore.getCanonicalServiceName();
   }
 
+  @VisibleForTesting
+  public void setClock(Clock monotonicClock) {
```
Comment (fapifta) [inline hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneClientAdapterImpl.java:441]: I think we can reduce the visibility to default, as tests that should tinker with the clock can be - and for now they are - in the same package. What do you think?

**Review by fapifta — COMMENTED** [review]
Hi @umamaheswararao thank you for working on this after the long discussions about the problem.
The solutions seems to be good, I have left two minor comments inline, otherwise I am +1 for the rest.

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/OzoneClientUtils.java [hunk 3]
```diff
@@ -64,4 +73,78 @@ public static BucketLayout resolveLinkBucketLayout(OzoneBucket bucket,
...
+   *
+   * @param replication                - replication value passed from FS API.
+   * @param clientConfiguredReplConfig - Client side configured replication
+   *                                   config.
+   * @param bucketReplConfig           - server side bucket default replication
+   *                                  config.
+   * @param config                     - Ozone configuration object.
+   * @return client resolved replication config.
+   */
+  public static ReplicationConfig resolveClientSideReplicationConfig(
```
Comment (sodonnel) [inline hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/OzoneClientUtils.java:89]: Should we add a unit test for this logic? There is a fair bit going on here to not have a test covering it.

**Comment by sodonnel** [comment]
This change looks mostly good to me. I have a couple of comments above.

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneClientAdapterImpl.java (hunk 1 above)
Comment (umamaheswararao) [reply hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneClientAdapterImpl.java]: Fixed it. Yeah unnecessary format.

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneClientAdapterImpl.java (hunk 2 above)
Comment (umamaheswararao) [reply hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneClientAdapterImpl.java]: done

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/OzoneClientUtils.java (hunk 3 above)
Comment (umamaheswararao) [reply hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/OzoneClientUtils.java:89]: Added few tests specific to this API

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneClientAdapterImpl.java [hunk 4]
```diff
@@ -211,29 +239,20 @@ public OzoneFSOutputStream createFile(String key, short replication,
...
-            ReplicationConfig.adjustReplication(
-                replConfig, replication, config);
-        ozoneOutputStream =
-            bucket.createFile(key, 0, customReplicationConfig, overWrite,
-                recursive);
-      } else {
-        ozoneOutputStream =
-            bucket.createFile(key, 0, replConfig, overWrite, recursive);
-      }
+      ReplicationConfig clientConfiguredReplConfig =
```
[the remark this answers is not in the document]
Comment (umamaheswararao) [reply hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneClientAdapterImpl.java]: did that. Thanks

**Comment by umamaheswararao** [comment]
Thanks @sodonnel and @fapifta for reviews. I have just push another commit to address the comments. please take a look. Thanks

**Review by sodonnel — APPROVED** [review]
This change LGTM

**Pull request merged**